parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+                    return Table(return_list[0:limit])
+
+    def get_campaign_tags(self, campaign=None, limit=None, per_page=25, filter=None):
+        """
+        Retrieve all tags (i.e. custom field values) within provided limit and filters
+        `Args:`
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+            limit: int
+                The number of entries to return. When None, returns all entries.
+            per_page: int
+                The number of entries per page to return. 25 maximum and default.
+            filter
+                The OData query for filtering results. E.g. "modified_date gt '2014-03-25'".
+                When None, no filter is applied.
+        `Returns:`
+            Parsons Table of full set of tags available in Action Builder.
+        """
+
+        return self._get_entry_list(
+            campaign, "tags", limit=limit, per_page=per_page, filter=filter
+        )
+
+    def get_tag_by_name(self, tag_name, campaign=None):
+        """
+        Convenience method to retrieve data on a single tag by its name/value
+        `Args:`
+            tag_name: str
+                The value of the tag to search for.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Parsons Table of data found on tag in Action Builder from searching by name.
+        """
+
+        filter = f"name eq '{tag_name}'"
+
+        return self.get_campaign_tags(campaign=campaign, filter=filter)
+
+    def insert_new_tag(self, tag_name, tag_field, tag_section, campaign=None):
+        """
+        Load a new tag value into Action Builder. Required before applying the value to any entity
+        records.
+        `Args:`
+            tag_name: str
+                The name of the new tag, i.e. the custom field value.
+            tag_field: str
+                The name of the tag category, i.e. the custom field name.
+            tag_section: str
+                The name of the tag section, i.e. the custom field group name.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Dict containing Action Builder tag data.
+        """
+
+        campaign = self._campaign_check(campaign)
+        url = f"campaigns/{campaign}/tags"
+
+        data = {
+            "name": tag_name,
+            "action_builder:field": tag_field,
+            "action_builder:section": tag_section,
+        }
+
+        return self.api.post_request(url=url, data=json.dumps(data))
+
+    def upsert_entity(
```

Comment:
My instinct is to make this an internal function and have the public api specify which entity is being updated. Might make the function validation simpler as well

Reply:
I agree, though perhaps a little grudgingly. Both Action Network and Action Builder use a common endpoint for inserting and updating: the signup helper. Action Network has much simpler upsert logic, because records are unique on additional data: email and/or mobile number. Action Builder has no such uniqueness, so you're right that these should be separate. I'm just grumbly because I need to make sure this change is reflected in the sync that's live, but that's on me for running a sync on branch like that, before the connector is even merged.

Reply:
I could also see an argument for both. Having convenience methods like `upsert_person` (or whatever it would be) that are just syntactic sugar on the more general `upsert_entity`. But I do think having the more semantically clear parts of the API is important

Reply:
oh interesting. we may have missed each other. I think upsert is a better public api to use than an insert/update split. I was suggesting the _person, _campaign versions. 

we might be reaching the limits of asyncronous review and may need to grab some time